### PR TITLE
feat(dbal): support DBAL 4.x

### PR DESCRIPTION
**Pull request by tcitworld**

### CODE REVIEW

**Comment by sunchaserinfo** [comment]
nice, thanks

File: src/CidrType.php [hunk 1]
```diff
@@ -39,11 +39,11 @@ public function convertToDatabaseValue(mixed $value, AbstractPlatform $platform)
...
-            default => throw ConversionException::conversionFailedInvalidType(
-                $value,
-                $this->getName(),
-                ['null', IPBlock::class]
-            ),
+            default => throw new ConversionException(sprintf('Could not convert PHP value %s to type %s. Expected one of the following types: %s',
+                var_export($value, true),
+                self::NAME,
+                implode(', ', ['null', IPBlock::class])
+            ))
```
Comment (sunchaserinfo) [inline src/CidrType.php]: please use InvalidType::new here
https://github.com/doctrine/dbal/blob/4.0.x/src/Types/Exception/InvalidType.php#L29

File: src/InetType.php [hunk 2]
```diff
@@ -41,11 +41,11 @@ public function convertToDatabaseValue(mixed $value, AbstractPlatform $platform)
...
-            default => throw ConversionException::conversionFailedInvalidType(
-                $value,
-                $this->getName(),
-                ['null', IP::class, IPBlock::class]
-            )
+            default => throw new ConversionException(sprintf('Could not convert PHP value %s to type %s. Expected one of the following types: %s',
+                var_export($value, true),
+                self::NAME,
+                implode(', ', ['null', IP::class, IPBlock::class])
+            ))
```
Comment (sunchaserinfo) [inline src/InetType.php]: as in CidrType

File: src/InetType.php [hunk 3]
```diff
@@ -28,7 +28,7 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): IP|
         try {
             return str_contains($value, '/') ? IPBlock::create($value) : IP::create($value);
         } catch (InvalidArgumentException $e) {
-            throw ConversionException::conversionFailed($value, $this->getName(), $e);
+            throw new ConversionException('Could not convert database value "' . $value . '" to Doctrine Type ' . self::NAME, 0, $e);
```
Comment (sunchaserinfo) [inline src/InetType.php]: as in CidrType

File: src/CidrType.php [hunk 4]
```diff
@@ -27,7 +27,7 @@ public function convertToPHPValue(mixed $value, AbstractPlatform $platform): ?IP
         try {
             return IPBlock::create($value);
         } catch (InvalidArgumentException $e) {
-            throw ConversionException::conversionFailed($value, $this->getName(), $e);
+            throw new ConversionException('Could not convert database value "' . $value . '" to Doctrine Type ' . self::NAME, 0, $e);
```
Comment (sunchaserinfo) [inline src/CidrType.php]: Not 100% sure but let's go with
ValueNotConvertible::new
https://github.com/doctrine/dbal/blob/4.0.x/src/Types/Exception/ValueNotConvertible.php#L22

**Review by sunchaserinfo — CHANGES_REQUESTED** [review]

**Comment by sunchaserinfo** [comment]
Comments got slightly out of order, sry

**Comment by sunchaserinfo** [comment]
Generally, I want more specific exceptions

File: composer.json [hunk 5]
```diff
@@ -20,7 +20,7 @@
     },
     "require": {
         "php": "^8.0",
```
Comment (sunchaserinfo) [inline composer.json]: DBAL 4 requires PHP 8.1, please update here as well

**Review by sunchaserinfo — CHANGES_REQUESTED** [review]

**Comment by tcitworld** [comment]
> Generally, I want more specific exceptions

Yeah, sorry about that, should have bothered to look a little bit further for replacements.

File: src/CidrType.php [hunk 6]
```diff
@@ -39,11 +40,7 @@ public function convertToDatabaseValue(mixed $value, AbstractPlatform $platform)
         return match (true) {
             $value === null => null,
             $value instanceof IPBlock => $value->withPrefixLength(),
-            default => throw ConversionException::conversionFailedInvalidType(
-                $value,
-                $this->getName(),
-                ['null', IPBlock::class]
-            ),
+            default => throw throw InvalidType::new($value, self::NAME, ['null', IPBlock::class])
```
Comment (sunchaserinfo) [inline src/CidrType.php]: You have 2 throws now :D

File: src/InetType.php [hunk 7]
```diff
@@ -41,11 +42,7 @@ public function convertToDatabaseValue(mixed $value, AbstractPlatform $platform)
             $value === null => null,
             $value instanceof IP => $value->humanReadable(),
             $value instanceof IPBlock => $value->withPrefixLength(),
-            default => throw ConversionException::conversionFailedInvalidType(
-                $value,
-                $this->getName(),
-                ['null', IP::class, IPBlock::class]
-            )
+            default => throw throw InvalidType::new($value, self::NAME, ['null', IP::class, IPBlock::class])
```
Comment (sunchaserinfo) [inline src/InetType.php]: here too

**Review by sunchaserinfo — CHANGES_REQUESTED** [review]

File: src/CidrType.php (hunk 6 above)
Comment (tcitworld) [reply src/CidrType.php]: :see_no_evil:

**Comment by sunchaserinfo** [comment]
lgtm

**Comment by sunchaserinfo** [comment]
released as 3.0.0

**Pull request merged**